src/zmalloc.c

```diff
@@ -88,10 +88,44 @@ void zlibc_free(void *ptr) {
 #define dallocx(ptr, flags) je_dallocx(ptr, flags)
 #endif
 
-#define update_zmalloc_stat_alloc(__n) atomic_fetch_add_explicit(&used_memory, (__n), memory_order_relaxed)
-#define update_zmalloc_stat_free(__n) atomic_fetch_sub_explicit(&used_memory, (__n), memory_order_relaxed)
+#if __STDC_NO_THREADS__
+#define thread_local __thread
+#else
+#include <threads.h>
+#endif
+
+/* A thread-local storage which keep the current thread's index in the used_memory_thread array. */
+static thread_local int thread_index = -1;
+/* Element in used_memory_thread array should only be written by a single thread which
+ * distinguished by the thread-local storage thread_index. But when an element in
+ * used_memory_thread array was written, it could be read by another thread simultaneously,
+ * the reader will see the inconsistency memory on non x86 architecture potentially.
+ * For the ARM and PowerPC platform, we can solve this issue by make the memory aligned.
+ * For the other architecture, lets fall back to the atomic operation to keep safe. */
+#if defined(__i386__) || defined(__x86_64__) || defined(__amd64__) || defined(__POWERPC__) || defined(__arm__) ||      \
+    defined(__arm64__)
+static __attribute__((aligned(sizeof(size_t)))) size_t used_memory_thread[MAX_THREADS_NUM];
+#else
+static _Atomic size_t used_memory_thread[MAX_THREADS_NUM];
+#endif
+static atomic_int total_active_threads;
 
-static _Atomic size_t used_memory = 0;
+/* Register the thread index in start_routine. */
+static inline void zmalloc_register_thread_index(void) {
+    thread_index = atomic_fetch_add_explicit(&total_active_threads, 1, memory_order_relaxed);
```

Comment:
Good point. It's not very good to assume that.

I wonder if we can let this index overflow to zero. Worst case, if two threads use the same index, the implementation is still correct, right?

If we can do that, then zmalloc doesn't need to know the server's exact max number of threads. I prefer if we can avoid that coupling.

Reply:
>Good point. It's not very good to assume that.

+1.

I don't think the engine dynamically creates/destroys threads today. that said, agreed we could start doing that in the future so there is indeed a risk. moreover, I wonder if this assumption could be broken already by a module allocating/freeing memory from dynamically created and short-lived threads.

Reply:
what about `HelloBlock_ValkeyCommand` [didn't look at the logic of it TBH]? In general what about modules?

Reply:
BTW, my own preference is to wrap the thread create function and plug any initialization we want to it. I can imagine other usages for that, I guess you discussed this and I'm late to the party

Reply:
@PingXie didn't read carefully enough to see you mentioned the module part, sorry

Reply:
Why complicate the matter? We can just wrap the thread index modulo MAX_THREADS_NUM. The implementation is still correct. In the worst case, we'll just have two threads share the same variable, so we get slightly less out of this optimization.

Pros:

* zmalloc doesn't need to know the real max number of threads. De-coupling that.

* We can delete this todo:

  ```C
      /* TODO: Handle the case when exceed the MAX_THREADS_NUM (may rarely happen). */
      assert(total_active_threads < MAX_THREADS_NUM);
  ```

Am I missing anything?

Reply:
OK, I understand now why two threads cant `add` to the same variable: `add` is not atomic. It is read-modify-write. A simple read or write to one word is atomic, so that's why we can avoid atomic when we have only one writer. (@lipzhu explain to me in a chat.)

Reply:
Another idea, a simple protection: If thread index > MAX_THREADS_NUM, just use one special shared atomic variable for those threads, something like `_Atomic size_t used_memory_for_additional_threads`. These last threads will not benefit from the optimization, but it's a protection. It's used only if some modules create a lot of threads.

Reply:
So in case a module or a future dynamic io-threads implementation which spawns and kills threads all the time will converge to use atomic? We will probably need to address this limitation in our future work then.
I would like to ask why is that so bad to keep a "free-ids-list" and use a mutex to guard it when we allocate thread id? It is a once in a thread lifecycle operation so basically I am not sure will cause that much performance impact. 
For the deregistration when a thread is destroyed. I would suggest think of using pthread_key_t so we can assign a destructor call to return the id.

Reply:
Yes, the module threads will use atomic in that case, but the main thread and IO threads use their own variable if they're spawned first and stay alive. We can improve this when/if we see a problem with modules. (Lipeng's idea is to keep a bitmap to flag each entry in the array as used or free.)

Reply:
I think this is fine. server threads are created earlier and long-lived. I don't think they will ever use the overflow atomic counter.